assets/src/models/vehicle.ts

```diff
@@ -1,19 +1,26 @@
-import { Ghost, Vehicle, VehicleOrGhost } from "../realtime"
+import { Ghost, Vehicle, VehicleInScheduledService } from "../realtime"
 import { Route } from "../schedule"
 import { now } from "../util/dateTime"
 
 export const isVehicle = (
-  vehicleOrGhost: VehicleOrGhost
+  vehicleOrGhost: Vehicle | Ghost
 ): vehicleOrGhost is Vehicle => !isGhost(vehicleOrGhost)
 
+export const isVehicleInScheduledService = (
+  vehicleOrGhost: Vehicle | Ghost
+): vehicleOrGhost is VehicleInScheduledService =>
+  !isGhost(vehicleOrGhost) && vehicleOrGhost.directionId !== null
```

Comment:
The main reason things like route and trip would be missing are for invalid vehicles (one of the reasons I think it would be good to split those out into their own type at some point). For operator info, some of that data comes specifically from BusLoc so I don't want to fail to display vehicles if we have a partial data outage but still have information from Swiftly.

Reply:
Ah that makes sense to me for trip id & operator id! I think routeId would be present though since we show invalid vehicles on the route ladder?

Reply:
It looks like I made it nullable for vehicles in general as part of handling shuttles (629cf179fd0ab4033c8723b658d742fb72c49191) and then never added a constraint that vehicles in scheduled service need to have non-null route IDs. I'm sort of torn on this - if it typechecks it means that the various places that can display vehicles in scheduled service have some default handling of route ID to take care of the case where it's null, so why make the type stricter? But of course in our data model it does seem like it should always be filled in. I think for the time being it's safer to leave it nullable personally, even if in some ways I know that's not a super satisfying explanation.

Reply:
> why make the type stricter?

I realize this may not sway your stance, but just to share more of some of the reasoning behind my original comments

* As far as I can tell, everywhere we are fetching data for `VehicleInScheduledService` is fetching by routeId, so I would expect routeId to need to be populated. 

* If I'm reading this guard as a sentence, "A vehicle that isn't a ghost and has a direction is in scheduled service" doesn't seem to tell the whole story. Thinking about my understanding of a vehicle in scheduled service outside of code, I would describe it as "a vehicle assigned to a run & a route that is not a shuttle is in scheduled service". There are some other fields in here that weren't brought up in my earlier comments (run & shuttle) which may be incorrect assumptions, but I having a more robust typecheck could help readers to understand what it means to be in scheduled service.
* I also have an assumption (which may be incorrect) that the fields of route_id and direction_id are tied together - what does direction_id mean without a route_id? 

I'd prefer to avoid making routeId nullable here if we can avoid it so that we don't have to add/maintain logic to handle that case if we know it shouldn't be hit.

Reply:
> As far as I can tell, everywhere we are fetching data for VehicleInScheduledService is fetching by routeId, so I would expect routeId to need to be populated.

Okay, this is a good point.

> If I'm reading this guard as a sentence, "A vehicle that isn't a ghost and has a direction is in scheduled service" doesn't seem to tell the whole story. Thinking about my understanding of a vehicle in scheduled service outside of code, I would describe it as "a vehicle assigned to a run & a route that is not a shuttle is in scheduled service". There are some other fields in here that weren't brought up in my earlier comments (run & shuttle) which may be incorrect assumptions, but I having a more robust typecheck could help readers to understand what it means to be in scheduled service.

I think in addition to the question about `route_id` itself this gets to some questions about user-defined type predicates in TypeScript. The way the implementation works, it actually didn't check much of anything about the structure of the data itself (for instance the `isGhost` check that looks at the contents of the ID field but nothing else), and TypeScript doesn't actually enforce anything about how your type predicates work. So I'm not sure of the best way to approach this is. One option is using `instanceof` but these aren't actually classes so it would probably require a lot more work.

Anyhow, in any event I'll make route_id at least non-nullable again.

Reply:
8a5c8683